Bounty: Mazda

Jafaral opened the pull request:
~There is still work to be done but I think a PR gives more visibility to this port with more potential help from members of the community~. I have been working on this port  since late last year with help from a few members in the community (extra thanks to @bugsy924).

~~Unfortunately, Mazda has the same steer lockout issue as ford~~ 
[update Nov 2019: The lockout in Mazda is different from Ford. Ford lockout happens at regular intervals and there is no solution for that. Mazda on the other hand just forces the driver to keep hands on the steering all the time with a grace period of only 5 seconds with no driver torque. If driver torque is detected, the LKAS system functions normally with no issue. Applying a weight to the steering fixes the issue completely. ]

If the hands are off the steering wheel for more than 5 seconds you get a warning and the steer commands eventually get ignored. So far I have not found a solution for that,  maybe we will have better luck with new models next year.   My solution until now is to use a weight on the steering wheel. 

Torque is modest but enough to handle most curves on a freeway (up to 8-10 degrees). I've driven 10s of hours so far with no major issues. A few users in the community have been using my port and they reported positive results.

~I need to work more on the safety side of things, this will be done in the next few weeks. I have put all of my energy so far to find a programmatic solution for the lock out issue.~ [DONE]

The Mazda giraffe is available here:
https://github.com/Jafaral/neo/tree/mazda/giraffe/mazda

~I will submit a PR against the neo repo, but I need to move the lateral control to CAN 2, in the current design that is on CAN 1.~

[Update Nov 2019] : The giraffe now has lateral control on CAN 2, although it doesn't matter with the new black panda.  DBC files and panda safety were all upstreamed!

Comments are welcome!
Jafar/Jacar

Jafaral commented:
@rbiasini Should I rebase my changes on top of devel and squash most of the changes ?  guess I also have to remove the dbc and panda files since those should go into their own repos. I already upstreamed the dbc file but I made some tweaks to it.

rbiasini commented:
@Jafaral, thanks for opening this PR. Looks like the overall architecture is good. Few preliminary comment:
- safety code needs work
- the 10s lockout unfortunately would make the port unmergeable :(. 
- as far as CAN enumeration: yes, you should swap CAN 1 and CAN 2.

bugsy924 commented:
would a read-only version be upstreamable with a link to a community maintained openpilot fork?

On Jul 31, 2019 1:05 PM, rbiasini <notifications@github.com> wrote:

@Jafaral<https://github.com/Jafaral>, thanks for opening this PR. Looks like the overall architecture is good. Few preliminary comment:

  *   safety code needs work
  *   the 10s lockout unfortunately would make the port unmergeable :(.
  *   as far as CAN enumeration: yes, you should swap CAN 1 and CAN 2.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/commaai/openpilot/pull/762?email_source=notifications&email_token=AHMDMSF2PEUQMGGVXLL6MS3QCD6PDA5CNFSM4IHAAQTKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3F5LXI#issuecomment-516675037>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AHMDMSF2G5H2N4U57B7DLQDQCD6PDANCNFSM4IHAAQTA>.

rbiasini commented:
@bugsy924 actually yes. Same thing we do with Ford: we just won't upstream `carcontroller.py` and `mazdacan.py`.

Jafaral commented:
@bugsy924 great suggestion. @rbiasini thanks for the feedback. I will work on getting a read-only version cleaned up for upstream merge.

Jafaral commented:
https://www.youtube.com/watch?v=qYF_1PSZWdQ

Jafaral commented:
@rbiasini what if I make carcontroller.py and mazdacan.py themselves read-only,  no control packets are sent, and the panda also would be in read-only mode, can we upstream it that way? That way the port would have the complete logic but without any actual control. I have done more than 30 hours driving over the last few days and pretty comfortable moving forward with this. I will post more videos if needed.

Jafaral commented:
Two lanes highway drive:

https://www.youtube.com/watch?v=f0MayhGbmmY

avion540 commented:
@Jafaral I have a 2019 Mazda 3 Hatchback and I don't believe I have the 5 second auto lockout. I do not get the "Please hold steering wheel" prompt. So is it possible that Mazda has removed that feature? 
Is there a way I can contact you privately (so we don't clog up this pull request) to find out if this is actually the case?

Jafaral commented:
@avion540 another Mazda3 user reported the same thing - no warning after 5 seconds. This is really promising news. We need to confirm this by running OP on one of those new cars to confirm. Join the discussion on comma's discord server, we have a Mazda channel and also you can DM me there @Jacar

Jafaral commented:
With the dbc and panda upstreamed I went a head dropped those files from this. I also squashed all commits to two, one for the simple boardd mazda safety addition and the other is the the whole new mazda directory. I also rebased on top of master. Will be testing this more over the next few days.

rbiasini commented on selfdrive/car/mazda/carstate.py:
```diff
@@ -0,0 +1,235 @@
+    # initialize can parser
+    self.CP = CP
+    #self.CAM_LT = CAM_LaneTrack(0, -1, 0, 0, 0, 0, 0, 0, 0)
+    self.CAM_LKAS = CAM_LaneKAS(0, -1, 0, 0, 0, 0, 0, 0)
```
the init method isn't really required. Just make sure you set the values the first time you call the car state update by simply setting the ctr to -1.

rbiasini replied:
Actually, the classes CAM_LaneTrack, CAM_LaneKAS, STEER_LKAS don't need to be defined.
See how it's handled in selfdrive/car/hyundai for example

Jafaral replied:
That was early in the development. I then found out the simpler way, but it was a case of  "if it in't broken, don't fix it" I stopped using most of these msgs, so better to just drop em. will do that.

rbiasini commented on selfdrive/car/mazda/carcontroller.py:
```diff
@@ -0,0 +1,91 @@
+    self.STEER_STEP = 1                  # how often we update the steer cmd
+    self.STEER_DELTA_UP = 10             # torque increase per refresh
+    self.STEER_DELTA_DOWN = 20           # torque decrease per refresh
+    if car_fingerprint == CAR.CX5:
```
not needed if

Jafaral replied:
I left it there just to make it clear for the next guy who port a new mazda to tweak it, although all current gen Mazda seems to be the same. The new Mazda3 could be different.

rbiasini commented on selfdrive/car/mazda/carstate.py:
```diff
@@ -0,0 +1,235 @@
+
+  def update(self, pt_cp, cam_cp):
+
+    self.can_valid = pt_cp.can_valid
```
unused I think

rbiasini commented on selfdrive/car/mazda/carstate.py:
```diff
@@ -0,0 +1,235 @@
+
+    self.cam_laneinfo = cam_cp.vl["CAM_LANEINFO"]
+
+    if self.CAM_LKAS.ctr != cam_cp.vl["CAM_LKAS"]['CTR']:
```
why you can't simply do:
```
self.CAM_LKAS = cam_cp.vl["CAM_LKAS"]
```
regardless the counter value?

Jafaral replied:
Again, in early development I couldn't get the car to steer, so I had code at some to sync with the stock LKAS ctr. Will drop.

rbiasini commented on selfdrive/car/mazda/carstate.py:
```diff
@@ -0,0 +1,235 @@
+    self.sig3 = s3
+    self.chksum = ck
+
+class CAM_LaneKAS():
```
unused?

rbiasini commented on selfdrive/car/mazda/carstate.py:
```diff
@@ -0,0 +1,235 @@
+
+  return CANParser(DBC[CP.carFingerprint]['pt'], signals, checks, canbus.cam)
+
+class CAM_LaneTrack():
```
unused?

Jafaral replied:
will drop!

rbiasini commented on selfdrive/car/mazda/carcontroller.py:
```diff
@@ -0,0 +1,91 @@
+      else:
+        line_not_visible = 1
+
+      e1 = 0 #CS.CAM_LKAS.err1
```
intended to be ignored?

Jafaral replied:
I used to pass error codes, not necessary and not needed anymore. will drop it

rbiasini commented on selfdrive/car/mazda/carcontroller.py:
```diff
@@ -0,0 +1,91 @@
+
+class CarControllerParams():
+  def __init__(self, car_fingerprint):
+    self.STEER_MAX = 600                 # max_steer 2048
```
woha, so 2048 must be a lot. How did you pick 600?

bugsy924 replied:
probably a leftover from subaru

Jafaral replied:
Not really, the torque signal is unsigned 12 bit with offset value of 2048 so really the allowed values are [-2048:2047]. I tested values in all of this range and they all worked. The EPS sends out to signals in response to torque signal coming from the camera. Call these ESP_TORQUE which is a copy of the camera torque signal, and the other msg EPS_TORQUE_EFFECTIVE which is the value of the torque actually applied by the EPS. The actual value is capped at 308. i,e range is [-308:308]. The new signal is not a simple trim of the original signal . For example this would be always 0 if torque is not allowed regardless of the value coming from the source signal. It is scaled, smoothed, and trimmed otherwise. I haven't spent enough time to try and decipher the exact translation but I I got the best results  if I limit the max value in the range  of[500 :900] , I think I'm driving with 750 now. higher values seem to decrease steer response due to smoothing I think.  i.e it takes a little longer for the eps to smooth the values when there is huge jumps. I'm assuming Mazda did that to prevent jerky steering. In all cases the value 600 is a very good value. If someone wants to spend time reverse engineer the EPS logic and come up with a better value I'd be happy to use it ;p

Jafaral commented:
@rbiasini Thanks for taking the time to review this. I will  drop the unused or otherwise the nice to have features but not necessary for the port. I came to the same conclusion myself. What I drive with now is the simplest code path, no changes to any of the dash msgs, everything is stock. The only thing I send  is the LKAS msg. I don't even drop the stock LKAS and the car is just happy that way. since I send LKAS at a much higher rate than stock the EPS just ignores the stock LKAS msgs and uses what is coming from OP. This allows me to run OP with zero chances of dashboard errors, I haven't seen one in a long time. I can plug the EON anytime even when the car is driving and unplug it any time even when the car is driving in everything works smoothly. The result is zero friction with EON. I don't have to take it off overnight and keep it charged. If it is dead in the morning, but the time I drive a mile or so, it is booted and running as it should without any intervention! I love it!

rbiasini commented on selfdrive/car/mazda/interface.py:
```diff
@@ -0,0 +1,217 @@
+    self.cam_cp = get_cam_can_parser(CP, canbus)
+
+    self.CC = None
+    if CarController is not None:
```
Not needed?

Jafaral replied:
Which part you are referring to exactly?

Jafaral replied:
Oh u mean no need for "if"?

rbiasini replied:
nevermind, my bad. it's present in all car interfaces. I'll revisit this more generally.

rbiasini commented on selfdrive/car/mazda/interface.py:
```diff
@@ -0,0 +1,217 @@
+    
+    tire_stiffness_factor = 0.70   # not optimized yet
+
+    if candidate in [CAR.CX5]:
```
Not needed

Jafaral replied:
What is not needed?

Jafaral replied:
I assume same as above,  u r referring to "if"

rbiasini replied:
sorry, my bad here too. The `if car...` here is fine.

geohot commented:
If this pull request is still relevant, please resubmit it to the master branch. This should allow our development flow to work better, we commit to master, you commit to master, everyone commits to master! Learn more: https://medium.com/@comma_ai/a-2020-theme-externalization-13b33326d8b3
